internal/cfgmgr/cfgmgr.go

```diff
@@ -0,0 +1,163 @@
+
+type devPropKey struct {
+	fmtid guid.GUID
+	pid   uint32
+}
+
+const (
+	_CM_GETIDLIST_FILTER_BUSRELATIONS = uint32(0x00000020)
+
+	_CM_LOCATE_DEVNODE_NORMAL = uint32(0x00000000)
+
+	_DEVPROP_TYPE_STRING        = uint32(0x00000012)
+	_DEVPROP_TYPEMOD_LIST       = uint32(0x00002000)
+	_DEVPROP_TYPE_STRING_LIST   = uint32(_DEVPROP_TYPE_STRING | _DEVPROP_TYPEMOD_LIST)
+	_DEVPKEY_LOCATIONPATHS_GUID = "a45c254e-df1c-4efd-8020-67d146a850e0"
+)
+
+func getDevPKeyDeviceLocationPaths() (*devPropKey, error) {
+	guid, err := guid.FromString(_DEVPKEY_LOCATIONPATHS_GUID)
+	if err != nil {
+		return nil, err
+	}
+	return &devPropKey{
+		fmtid: guid,
+		// pid value is defined internally in devpkey.w
+		pid: 37,
+	}, nil
+}
+
+func GetDeviceLocationPathsFromIDs(ids []string) ([]string, error) {
+	result := []string{}
+	devPKeyDeviceLocationPaths, err := getDevPKeyDeviceLocationPaths()
+	if err != nil {
+		return nil, err
+	}
+	for _, id := range ids {
+		var devNodeInst uint32
+		err = cmLocateDevNode(&devNodeInst, id, _CM_LOCATE_DEVNODE_NORMAL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to locate device node for %s", id)
+		}
+		propertyType := uint32(0)
+		propertyBufferSize := uint32(0)
+
+		// get the size of the property buffer by querying with a nil buffer and zeroed propertyBufferSize
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, nil, &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get property buffer size of devnode query for %s with", id)
+		}
+
+		// get the property with the resulting propertyBufferSize
+		tempPropertyBufSize := propertyBufferSize
+		propertyBuffer := make([]uint16, propertyBufferSize*2)
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, &propertyBuffer[0], &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get location path property from device node for %s with", id)
+		}
+		if propertyType != _DEVPROP_TYPE_STRING_LIST {
+			return nil, fmt.Errorf("expected to return property type DEVPROP_TYPE_STRING_LIST %d, instead got %d", _DEVPROP_TYPE_STRING_LIST, propertyType)
+		}
+		if propertyBufferSize > tempPropertyBufSize {
+			return nil, fmt.Errorf("location path is too large for the buffer, size %d", propertyBufferSize)
+		}
+		formattedResult, err := convertFirstNullTerminatedValueToString(propertyBuffer[:propertyBufferSize])
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, formattedResult)
+	}
+
+	return result, nil
+}
+
+// helper function that finds the first \u0000 rune and returns the wide string as a regular go string
+func convertFirstNullTerminatedValueToString(buf []uint16) (string, error) {
+	r := utf16.Decode(buf)
+	converted := string(r)
+	zerosIndex := strings.IndexRune(converted, '\u0000')
+	if zerosIndex != -1 {
+		return "", fmt.Errorf("cannot convert value to string, malformed data passed: %v", buf)
```

Comment:
Not sure we should include the buffer in the error message, as it could be long.